phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -5181,6 +5192,24 @@ private void flushTable(byte[] tableName) throws SQLException {
         }
     }
 
+    @Override
+    public void refreshLiveRegionServers() throws SQLException {
+        synchronized (liveRegionServersLock) {
+            try (Admin admin = getAdmin()) {
+                this.liveRegionServers = new ArrayList<>(admin.getRegionServers(true));
+            }
+            catch (IOException e) {
+                throw ServerUtil.parseServerException(e);
+            }
+        }
+        LOGGER.info("Refreshed list of live region servers.");
+    }
+
+    @Override
+    public List<ServerName> getLiveRegionServers() {
+        return this.liveRegionServers;
```

Comment:
Do we want to have synchronize block here also? Assume a case where refreshing the live regionservers is ongoing and some client tries to access liveRegionServers. That client will get stale copy of region servers.

Reply:
@shahrs87 I think a client getting a stale copy of region server list should be okay. It will pick another region server randomly which will likely be different from the regionserver which had issues. Also the list variable is `volatile` so that all threads will get the latest updated value. This weaker and light-weight form of synchronization should suffice for our purpose, most of the shared state in CQSI is handled this way. Let me know if I am missing something.